proposals/multiple_user_script_worlds.md

```diff
@@ -0,0 +1,278 @@
+
+     /**
+      * Specifies which pages this user script will be injected into. See
+      * Match Patterns for more details on the syntax of these strings. This
+      * property must be specified for ${ref:register}.
+      */
+     matches?: string[];
+
+     /**
+      * Specifies when JavaScript files are injected into the web page. The
+      * preferred and default value is document_idle
+      */
+     runAt?: RunAt;
+
+     /**
+      * The JavaScript execution environment to run the script in. The default
+      * is `USER_SCRIPT`
+      */
+     world?: ExecutionWorld;
+
++    /**
++     * If specified, specifies a specific user script world ID to execute in.
++     * Only valid if `world` is omitted or is `USER_SCRIPT`. If omitted, the
++     * script will execute in the default user script world.
++     * Values with leading underscores (`_`) are reserved.
++     */
++    worldId?: string;
+   }
+
+   ...
+
++   /**
++    * Removes the configuration for a given world.
++    */
++   export function removeWorld(worldId: string): Promise<void>;
++
++   /**
++    * Returns a promise that resolves to an array of the the configurations
++    * for all user script worlds.
++    */
++   export function getWorldConfigurations(): Promise<WorldProperties[]>;
+ }
+```
+
+Note that the signatures of the related functions, including `configureWorld()`,
+`register()`, and others are left unchanged.
+
+When the developer specifies a `RegisteredUserScript`, the browser will use a
+separate user script world when injecting the scripts into a document. If
+`worldId` is omitted, the default user script world will be used.
+
+Worlds may be configured via `userScripts.configureWorld()` by indicating the
+given `worldId`. User scripts injected into a world with the given `worldId`
+will have the associated properties from the world configuration. If a world
+does not have a corresponding configuration, it uses the default user script
+world properties.
+
+World configurations can be removed via the new `userScripts.removeWorld()`
+method.
+
+Additionally, `runtime.Port` and `runtime.MessageSender` will each be extended
+with a new, optional `userScriptWorldId` property that will be populated in the
+arguments passed to `runtime.onUserScriptConnect` and
+`runtime.onUserScriptMessage` if the message initiator is a non-default user
+script world. We do not need to populate a value for messages from the default
+user script world, since this is unambiguous given the distinction in the
+event listeners (`onUserScriptMessage` already indicates the message was from
+a user script).
+
+### Behavioral Notes
+
+#### World Persistence
+
+Configured worlds are persisted until the owning extension is updated (in order
+to align with the behavior of the rest of the `userScripts` API) or manually
+removes them via the new `userScripts.removeWorld()` method.
+
+#### World Limits
+
+User agents may place limits on both the number of registered worlds and the
```

Comment:
Should we really limit the number of worlds? What are the numbers you are thinking of?

A too low limit reduces the practical value of the multiple worlds concept, and user script managers are then still forced to implement script wrappers to avoid conflicts between user scripts.

A reasonably high limit reduces the impact for users with a few scripts, but punishes power users with lots of scripts. If a user is willing to install many scripts and considers the overhead acceptable, why would we stop them?

Reply:
Assuming the world's overhead is ~1MB, how about 256 worlds max?

Reply:
Short answer: yes, I strongly think we should limit the number of worlds.  In particular, I think this is true for the active number of worlds on a page.

The registration is relatively cheap, so I don't think we need to be _too_ limiting for total number of registered worlds (I was originally thinking ~100, but I think that one's more flexible).  The real cost is in the number of active worlds on pages.  Each of these results in a separate JS context which, depending on the size of the script, may (very) non-trivially increase the weight of a given document.  Then, given the number of documents that can be actively running, this can become _much_ heavier weight.  Every additional user script world is roughly equivalent to the cost of a whole other content script-injecting extension.  From @tophf 's suggestion of 256 worlds assuming a 1MB overhead, that's 256MB _per document_.  (Assuming that was meant as a suggestion for "active worlds on a page", rather than total registered worlds.  I think 256 total registered worlds is perhaps a bit high, but not unreasonable.)  I just opened cnn.com (my go-to for a "noisy" site), and there are 30 different frames there.  That means that, with 256 worlds at 1MB a piece, the user would have 7.5GB of memory used on _that one tab_ just from user scripts.  I think that's more than we'd allow in Chrome.

Now, since a lot of user scripts are only intended to run on specific pages, there are lots of cases when two scripts would never "collide" and result in multiple worlds on the same page, so I'm hopeful that most user script managers wouldn't run into this limit frequently.  But for all the scripts that inject on every page, I think there is a risk of major performance impact.

I also think the exact values here may be something left up to individual browser vendors to establish, as we've done with historical limitations with performance (in which case, Firefox may have no limit, Safari on iOS may establish a different limit than Safari on MacOS, etc).  But rather than fully leave out the expectations from the specification and then have an individual browser impose an intervention in ways that are unpredictable to the developer, I think it's worth describing what the behavior the developer can expect will be.

Reply:
Chrome/Firefox similarly limit the number of OS processes they use and they combine sites in one process if necessary. Arguably extensions could set the maximum number of the worlds or at least read it from the API so they can display the warning to the user and add a "virtualizer" to the userscript code, specifically `with (globalProxy) {....}`, which is undesirable as `with` incurs overhead.

FWIW, my 1MB estimate might be too high. What if it's just 100kB?

Reply:
I see 46kB for an extension `Window` in Chrome Devtools -> Memory -> Snapshot.

Reply:
Unfortunately, it's not _just_ the overhead of the JS items -- there's more that needs to happen to create a new isolated world (on the Chrome side, we need a new v8::Context, ScriptState, execution environment, etc).  The impact I'm seeing for a completely empty script injection is about 200KB per world when injecting into a very simple site (example.com).  Because the isolated world also has a copy of the DOM (the DOM is shared, but the JS objects for the DOM are not), this increases with the complexity of the page.

I am supportive of exposing the maximum number of worlds so that extensions can handle this appropriately (as @tophf called out, potentially showing user-visible UI or modifying their injections to be "semi-isolated" from one another).

Reply:
So what happens to a matched userscript in the case its separate world cannot be created? There should be some event or another method of detecting it.

Reply:
This is called out in the proposal -- it will fallback to using the default world.

I'd be happy to consider:
- Having an additional trait in WorldProperties like "shouldFallBackIfWorldLimitReached" (perhaps with a better name) that controls whether the script should fall back or just bail and not inject, and / or
- Having a new event like "onUserScriptCouldNotInjectBecauseWorldLimitReached" (definitely with a better name)

However, given the scope creep we're seeing in this proposal, I'd like to do those in a separate change.

Reply:
...because the userscript manager extension would need to run the userscript's function inside `with(windowProxy)`

Reply:
Okay, as long as it's not postponed to some distant future.

Reply:
I'm not resolving the thread yet because the expressed questions has not been resolved, but feel free to merge even if this thread is still open.

Consider calling out improvements in this area in "future work".

Reply:
An event is necessary anyway, because managers need it to set up the script's environment before execution of any matched userscript e.g. set `window[randomId]` with the secure communication channel name in the MAIN world right right before execution so that no JS code can run in-between. Something like chrome.userScripts.onExec with (scriptId, worldId) parameters and chrome.runtime.lastError inside the listener.

Reply:
Synchronous cross-world communication and code execution has been discussed before, including recently at the in-person WECG meetup. That will become a separate proposal, and does not need to be specific to user scripts.

Reply:
My point was that userscripts need an event right before execution, it's not the same as cross-world communication. This event can be used to report the world creation error.

Reply:
Same-world userscripts in the default world need it too to set up the internal bindings for `GM` stuff which is implemented in a common controller script. BTW we need chrome.userScripts.registerController for the code that runs before the first matched userscript in the tab and receives all matched userscript ids.

Reply:
Sounds like we're agreed we can discuss these events in a follow-up proposal, so I'll resolve this conversation for now.